HPCC-29610 Fix broken file spray when no DropZoneRestriction

wangkx opened the pull request:
1. When spraying a file with no DropZoneRestriction, the getAndValidateDropZone() may not return a dropzone plane. The readAndCheckSpraySourceReq() should validate the return and read the plane name if there is the return.
2. In the checkDZScopeAccessAndSetSpraySourceDFUFileSpec(), validate DZPathScopePermissions only if the dropzone plane name is found in the readAndCheckSpraySourceReq().
3. Create throwOrLogDropZoneLookUpError(): if not bare metal or DropZoneRestriction enabled, throw an exception; otherwise, log the error.
4. Create findDropZonePlaneName(): if a plane name is not found based on dropzone host and path, call throwOrLogDropZoneLookUpError().
5. In ESP FileSpray service, call the findDropZonePlaneName() to find a dropzone plane name. Validate DZPathScopePermissions only if the plane name is found.
6. In getDZPathScopePermissions(), if a dropzone plane is not found based on dropzone host and path, call throwOrLogDropZoneLookUpError().

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

wangkx commented:
@jakesmith This PR is for fixing the crash in ESP FileSpray when spraying a file and isDropZoneRestrictionEnabled = false 
(see https://github.com/hpcc-systems/HPCC-Platform/pull/17356#discussion_r1206805445). Please review.

wangkx commented:
@jakesmith I just saw the smoke test error and added the fix.

jakesmith reviewed (commented):
@wangkx - please see comments.

wangkx commented:
@jakesmith please review my changes and let me know if I miss anything.

wangkx commented:
I noticed one smoke test failure. But, I do not know why. In my local test environment, I tested the spray from an empty Dir. The empty logical file is created.

jakesmith commented on esp/smc/SMCLib/TpCommon.cpp:
```diff
@@ -179,31 +208,40 @@ static SecAccessFlags getDropZoneScopePermissions(IEspContext& context, const IP
+            dropZoneName, dropZonePath, context.queryUserId(), getSecAccessFlagName(access), getSecAccessFlagName(accessReq));
+}
+
+extern TPWRAPPER_API SecAccessFlags getDZFileScopePermissions(IEspContext& context, const char* dropZoneName, const char* dropZonePath)
```
can you change 'dropZonePath' to e.g. 'dropZoneFilePath' to make it clear it's not just a path?
so it is differentiated from getDZPathScopePermissions which should be passed a path (not a path+filename).
(and in header too)

Same for checkDZFileScopePermissions.
Also the errors should clarify what is the scope being rejected in this case, as it stands it will include the filename, but the error says 'Access DropZone Scope' on both, which will lead the user to think a scope is needed for the filename.

wangkx replied:
changed

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -2912,17 +2913,11 @@ bool CFileSprayEx::onFileList(IEspContext &context, IEspFileListRequest &req, IE
-                    dropZoneName, sPath.str(), context.queryUserId(), getSecAccessFlagName(permission));
+                dropZoneName = findDropZonePlaneName(netaddr, sPath, dzName);
+            if (!isEmptyString(dropZoneName))
+                checkDZPathScopePermissions(context, dropZoneName, sPath, SecAccess_Read);
```
before checking LDAP, I think it should 1st validate that the path is within the named dropzone.
If there's not an existing function that checks if a path is valid for a plane(?), one should be added to dautils.?pp

jakesmith replied:
maybe beyond the scope of this PR, but the above code and use of legacy method validateDropZonePath (lines 2935-2940) should be revised (code like validateDropZonePath should be deleted).

I think this code section should look something like this:
```
        else
        {
            StringBuffer dzName;
            if (isEmptyString(dropZoneName))
                dropZoneName = findDropZonePlaneName(netaddr, sPath, dzName);
            Owned<IPropertyTree> dropZone;
            if (!isEmptyString(dropZoneName)) // must be true, unless bare-metal and isDropZoneRestrictionEnabled()==false
            {
                dropZone.setown(getDropZonePlane(dropZoneName));
                if (!dropZone)
                    throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Unknown landing zone: %s", dropZoneName);
                if (!isPathInDropZone(dropZone, sPath))
                    throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Path '%s' is not valid for dropzone '%s'", sPath.str(), dropZoneName);

                checkDZPathScopePermissions(context, dropZoneName, sPath, SecAccess_Read);
            }

            StringArray hosts;
            if (!isEmptyString(netaddr))
            {
                if (dropZone && !isHostInPlane(dropZone, netaddr, false))
                    throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Host '%s' is not valid for dropzone '%s'", netaddr, dropZoneName);

                hosts.append(netaddr);
            }
            else
            {
                // NB: if no netaddr, a dropZone must have been specified
                getPlaneHosts(hosts, dropZone);
                if (!hosts.ordinality())
                    hosts.append("localhost");
            }

            ForEachItemIn(i, hosts)
            {
                const char* host = hosts.item(i);
                getPhysicalFiles(context, dropZoneName, host, sPath, fileNameMask, directoryOnly, files);
            }
        }
```

wangkx replied:
The validateDropZonePath is also used in ws_fsBinding.cpp. I guess I should do the same (if there is dropzoneName, calls isPathInDropZone()) there.

wangkx replied:
I made the changes. Works for bare-metal. In my local Docker Desktop, the code failed in the isHostInPlane(dropZone, netaddr, false): Host 'localhost' is not valid for dropzone 'mydropzone'

wangkx replied:
Inside the isHostInPlane(), if I change the line 62 to: 
return streq(host, "localhost") ? true : false;
it seems working. I am not sure we should add that change. @jakesmith please advise.

jakesmith replied:
>return streq(host, "localhost") ? true : false;

I don't think that doesn't sound right.

Where is "localhost" coming from in this context?

jakesmith commented on esp/smc/SMCLib/TpCommon.cpp:
```diff
@@ -119,6 +119,35 @@ extern TPWRAPPER_API bool matchNetAddressRequest(const char* netAddressReg, bool
     return streq(netAddressReg, tpMachine.getConfigNetaddress());
 }
 
+static void throwOrLogDropZoneLookUpError(int code, char const* format, ...)
```
clearer if part of findDropZonePlaneName (see comment in findDropZonePlaneName)

jakesmith reviewed (changes requested):
@wangkx - please see comments.

jakesmith reviewed (changes requested):
@wangkx - please see comments.

wangkx commented on esp/smc/SMCLib/TpCommon.cpp:
```diff
@@ -135,6 +135,19 @@ extern TPWRAPPER_API void throwOrLogDropZoneLookUpError(int code, char const* fo
+    if (plane)
+        planeName.append(plane->queryProp("@name"));
+    else
+        throwOrLogDropZoneLookUpError(ECLWATCH_INVALID_INPUT, "DropZone not found for host '%s' path '%s'.", host, path);
```
Changed

jakesmith replied:
NB: tackled in a separate PR (targeting master).

wangkx commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -2279,24 +2280,26 @@ void CFileSprayEx::checkDZScopeAccessAndSetSpraySourceDFUFileSpec(IEspContext &c
     {
-        //Both srcPlane and srcHost are validated in readAndCheckSpraySourceReq().
+        //The srcHost is validated in readAndCheckSpraySourceReq().
+        //If the srcPlane is found in the readAndCheckSpraySourceReq(), the DZFileScopePermissions
```
Changed

jakesmith commented on dali/base/dautils.cpp:
```diff
@@ -149,6 +149,11 @@ IPropertyTree * findDropZonePlane(const char * path, const char * host, bool ipM
 
+bool isPathInPlane(IPropertyTree *plane, const char *path)
+{
+    return !isEmptyString(path) && startsWith(path, plane->queryProp("@prefix"));
```
As it stands:

> false == isPathInPlance(plane, "")
> false == isPathInPlance(plane, nullptr)

Hopefully it would not be called with an empty path, but I think if it is, it makes more sense to return true.

wangkx replied:
Changed

jakesmith commented on esp/services/ws_fs/ws_fsBinding.cpp:
```diff
@@ -386,6 +387,23 @@ void CFileSpraySoapBindingEx::xsltTransform(const char* xml, const char* sheet,
 }
 #endif
 
+static void checkDropZoneInputAndAccess(IEspContext &context, StringBuffer& dropZoneName, const char* host, const char* path, SecAccessFlags accessReq)
```
minor: dropZoneName isn't used by any of the callers, so it would be clearer if it were a const char *

jakesmith replied:
See other comments - looks like this method should be used by ws_fsService.cpp too.

wangkx replied:
If this method is used by ws_fsService.cpp, the StringBuffer& dropZoneName is needed and the method should return the dropzone tree.

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -2916,28 +2916,36 @@ bool CFileSprayEx::onFileList(IEspContext &context, IEspFileListRequest &req, IE
+                if (!dropZone)
+                    throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Unknown landing zone: %s", dropZoneName);
+                if (!isPathInPlane(dropZone, sPath))
+                    throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Path '%s' is not valid for dropzone '%s'", sPath.str(), dropZoneName);
```
Also, now you've consolidated the checks into checkDropZoneInputAndAccess, it looks like this code is very similar and checkDropZoneInputAndAccess should be used here too.

wangkx replied:
> Where is "localhost" coming from in this context?

The "localhost" comes from:
static void gatherDropZoneMachines(IArrayOf<IEspTpMachine> & tpMachines, IPropertyTree & plane)
{
    const char * prefix = plane.queryProp("@prefix");
    if (plane.hasProp("hosts"))
        gatherDropZoneMachinesFromHosts(tpMachines, plane, prefix);
    else
        tpMachines.append(*createHostTpMachine("localhost", prefix));
}

> I don't think that doesn't sound right.

The issue is because the getPlaneHostGroup() returns nullptr for my local Docker Desktop:
<planes category="lz" name="mydropzone" prefix="/var/lib/HPCCSystems/mydropzone"/>
What should I do to avoid this issue?

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -2916,28 +2916,36 @@ bool CFileSprayEx::onFileList(IEspContext &context, IEspFileListRequest &req, IE
-                Owned<IPropertyTree> dropZone = getDropZonePlane(dropZoneName);
-                if (!dropZone)
-                    throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Unknown landing zone: %s", dropZoneName);
+                if (dropZone && !isHostInPlane(dropZone, netaddr, false))
```
>Also, now you've consolidated the checks into checkDropZoneInputAndAccess, it looks like this code is very similar and checkDropZoneInputAndAccess should be used here too.

ditto. I think a call to checkDropZoneInputAndAccess should be used to validate in a common way.

jakesmith reviewed (commented):
@wangkx - I think looks a lot cleaner now, a few more comments.

wangkx commented:
@jakesmith thank you for your advices today. Please review my changes.

jakesmith reviewed (commented):
@wangkx - a couple of minor comments, please address and squash.

Have you verified these changes do not cause problems when looking up dropzones by a hostname (that doesn't match a dropzone) with restrictions switched off?

wangkx commented:
> Have you verified these changes do not cause problems when looking up dropzones by a hostname (that doesn't match a dropzone) with restrictions switched off?

Yes, I did.

wangkx commented:
@jakesmith sorry for taking a long time (mainly because of the builds in both bare-metal and DD before re-test). Please review.

wangkx commented on esp/smc/SMCLib/TpCommon.cpp:
```diff
@@ -221,9 +218,35 @@ extern TPWRAPPER_API void validateDropZoneAccess(IEspContext& context, const cha
+            throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Unknown landing zone: %s", dropZoneName.str());
+        if (isContainerized() && host && strsame("localhost", host)) //ECLWatch may send "localhost" for dropzones without hosts.
+            host = nullptr;
+        if (!isEmptyString(host) && !isHostInPlane(dropZone, host, false))
```
changed

wangkx commented on esp/smc/SMCLib/TpCommon.cpp:
```diff
@@ -221,9 +218,35 @@ extern TPWRAPPER_API void validateDropZoneAccess(IEspContext& context, const cha
+        Owned<IPropertyTree> dropZone = getDropZonePlane(dropZoneName);
+        if (!dropZone)
+            throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Unknown landing zone: %s", dropZoneName.str());
+        if (isContainerized() && host && strsame("localhost", host)) //ECLWatch may send "localhost" for dropzones without hosts.
```
Removed.

wangkx committed:
HPCC-29610 Fix broken file spray when no DropZoneRestriction

1. When spraying a file with no DropZoneRestriction, the
getAndValidateDropZone() may not return a dropzone plane. The
readAndCheckSpraySourceReq() should validate the return and read the
plane name if there is the return.
2. In the checkDZScopeAccessAndSetSpraySourceDFUFileSpec(), validate
DZPathScopePermissions only if the dropzone plane name is found in
the readAndCheckSpraySourceReq().
3. Create throwOrLogDropZoneLookUpError(): if not bare metal or
DropZoneRestriction enabled, throw an exception; otherwise, log the
error.
4. Create findDropZonePlaneName(): if a plane name is not found based
on dropzone host and path, call throwOrLogDropZoneLookUpError().
5. In ESP FileSpray service, call the findDropZonePlaneName() to find
a dropzone plane name. Validate DZPathScopePermissions only if the
plane name is found.
6. In getDZPathScopePermissions(), if a dropzone plane is not found
based on dropzone host and path, call throwOrLogDropZoneLookUpError().

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented:
@jakesmith this PR has been revised/tested for 9.0.x based on your advice.

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -2279,24 +2292,29 @@ void CFileSprayEx::checkDZScopeAccessAndSetSpraySourceDFUFileSpec(IEspContext &c
-            if (permission < SecAccess_Read)
-                throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Access DropZone Scope %s %s not allowed for user %s (permission:%s). Read Access Required.",
-                    srcPlane, file, context.queryUserId(), getSecAccessFlagName(permission));
+                SecAccessFlags permission = getDZFileScopePermissions(context, srcPlane, file, srcHost);
```
NB: tackled in a separate PR (targeting master).

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -2912,17 +2930,16 @@ bool CFileSprayEx::onFileList(IEspContext &context, IEspFileListRequest &req, IE
-                dropZoneName = plane->queryProp("@name");
+                SecAccessFlags permission = getDZPathScopePermissions(context, dropZoneName, sPath, nullptr);
+                if (permission < SecAccess_Read)
+                    throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Access DropZone Scope %s %s not allowed for user %s (permission:%s). Read Access Required.",
```
NB: tackled in a separate PR (targeting master).

jakesmith commented on esp/smc/SMCLib/TpCommon.cpp:
```diff
@@ -186,7 +202,14 @@ extern TPWRAPPER_API SecAccessFlags getDZPathScopePermissions(IEspContext& conte
+        dropZone.setown(findDropZonePlane(dropZonePath, dropZoneHost, true, false));
+        if (!dropZone)
+        {
+            throwOrLogDropZoneLookUpError(ECLWATCH_INVALID_INPUT, "getDZPathScopePermissions(): DropZone %s not found.", dropZoneName);
```
NB: tackled in a separate PR (targeting master).

jakesmith reviewed (approved):
@wangkx - looks good.
Have you opened a new JIRA for the other changes targeting master? (can you link it to this JIRA for reference)

wangkx commented:
The new JIRA is opened: [PCC-29670](https://track.hpccsystems.com/browse/HPCC-29670)
Refactor and remove of legacy code for ESP FileSpray service
